### PR TITLE
T002 review

### PR DESCRIPTION
**Pull request by dominiquesydow**
# Details

* Talktorial ID: 002
* Title: Molecular filtering: ADME and lead-likeness criteria
* Original authors: Michele Wichmann, Mathias Wajnberg
* Reviewer(s): Dominique Sydow, Andrea Volkamer
* Date of review: 2020-09-28
* [ReviewNB link](https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/)

# Content review

* Potential labels or categories (e.g. machine learning, small molecules, online APIs): _small molecules, drug-likeness, ADME, Lipinski's rule of five, radar plot, hydrogen-bond donor, hydrogen-bond acceptor, LogP, molecular weight, physicochemical properties_
* One line summary: _The chemical structure of a molecule can be used to determine physicochemical properties and estimate drug-likeliness. An often used criteria for a molecule's solubility and permeability is the Lipinski's rule of five, often visualized in the form of radar plots._ 
* [x] The table of contents reflects the talktorial story-line; order of #, ##, ### headers is correct
* [x] URLs are linked with meaningful words, instead of pasting the URL directly or linking words like `here`.
* [x] I have spell-checked the notebook
* [x] Images have enough resolution to be rendered with quality, without being _too_ heavy.
* [x] All figures have a description
* [x] Markdown cell content is still in-line with code cell output (whenever results are discussed)
* [x] I have checked that cell outputs are not incredibly long (this applies also to `DataFrames`)
* [x] Formatting looks correctly on the Sphinx render (bold, italics, figure placing)

# Code review

* Time it took to execute (approx.): _A couple of seconds._
* [x] Variable and function names follow snake case rules (e.g. `a_variable_name` vs `aVariableName`)
* [x] Spacing follows PEP8 (run Black on the code cells if needed)
* [ ] Code line are under 99 characters each (run `black -l 99`)
* [x] Comments are useful and well placed
* [x] There are no unpythonic idioms like `for i in range(len(list))` (see slides)
* [x] All 3rd party dependencies are listed at the top of the notebook
* [x] I have marked all code cell with output referenced in markdown cells with the label `# TODO: CI`
* [x] I have identified potential candidates for a code refactor / useful functions
* [x] All `import ...` lines are at the top (practice part) cell, ordered by standard library / 3rd party packages / our own (`teachopencadd.*`)
* [x] I have update the relative paths to absolute paths.
* List here unfamiliar libraries you find in the imports and their intended use: _None._

### CODE REVIEW

File: teachopencadd/talktorials/002_compound_adme/talktorial.ipynb [hunk 1]
```diff
@@ -6,11 +6,11 @@
    "source": [
```
Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>add reviewer:</p><p>Andrea Volkamer, 2020,&nbsp;<a href="https://volkamerlab.org/" target="_blank">Volkamer lab</a>, Charité</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>I started using T002 and so on everywhere</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Add one more sentence for CADD context:  </p><p>In the context of drug design, it is important to filter candidate molecules by e.g. their physicochemical properties. In this talktorial, the compounds acquired from ChEMBL (<strong>Talktorial 001</strong>) will be filtered by Lipinsik's rule of five to keep only orally bioavailable compounds (applying Lipinski's rule of five).</p><p>-&gt; change </p><p>-- 'active' to 'bioavailable'</p><p>-- 'check for' to 'keep only'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <ul><li><a href="http://www.swissadme.ch/" target="_blank">SwissADME&nbsp;</a> webserver</li><li>Description of 'lead compounds' ...&nbsp;</li><li>What is LogP?&nbsp;...</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>we can predict -&gt; we aim to predict ?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <ul><li>Sometimes, ADME(T) also includes&nbsp;<strong>T</strong>oxicology. -&gt; Often, ADME also includes&nbsp;<strong>T</strong>oxicology and is thus referred to as ADMET or ADMETox.</li><li>The amount and the time drug-uptake ... -&gt; The amount and the time <u>of</u> drug-uptake ...</li><li>The distribution of an absorbed substance, i.e. within the body, between blood and different tissues, and crossing of the blood-brain barrier are affected by regional blood flow rates, molecular size and polarity of the compound, and binding to serum proteins and transporter enzymes. -&gt; The distribution of an absorbed substance, i.e. within the body, between blood and different tissues, <u>or</u> crossing <s>of</s> the blood-brain barrier are affected by regional blood flow rates, molecular size and polarity of the compound, <u>as well as</u> binding to serum proteins and transporter enzymes.</li><li>As soon as a compound enters the body, it usually starts to be metabolized. -&gt; Maybe: After entering the body, the compound will be metabolized.</li><li>Sentence? This way, toxic compounds can be removed from the body, however the metabolism itself can also Reducing the amount of an absorbed substance can be favorable if a toxic compound is removed. transformation of a chemical could even yield new toxic metabolites. </li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <ul><li>... to find desired drugs -&gt; to develop effective drugs</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>I suggest to move (or remove) this sentence 'In this tutorial, you learn how to generate a radar plot using the Python plotting library&nbsp;<code>matplotlib</code>.&nbsp;'</p><p>Because here it's more about the theoretical aspect.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>'We import the necessary libraries,&nbsp;'</p><p>This part could be remove here, because the libraries are imported two cells above already?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Add a markdown cell above:</p><p>'First, we combine the names and SMILES of the molecules, together with their structure, in a dataframe.'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>is it a descriptor library? otherwise maybe:</p><p>using&nbsp;some of the descriptors (link) available in rdkit</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>add comment above colors like</p><p>'# colors are used for plotting the molecules later'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>(not super urgent!) would it be possible to add a tick per molecule (1 and 3 are not showing)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>In the above bar chart we compared</p><p>-&gt; In the above bar chart, we compared</p><p> </p><p>... four example <strong>drug</strong> molecules ...</p><p>The four examples are note necessarily drugs or are they? Maybe just delete?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <ul><li>suggest to rename function: calculate_ro5 -&gt; compliance_to_ro5 or ro5_compliance or 	calculate_ro5_compliance?</li><li>one could also add the parameters and returns of the function</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>However, since all of them are approved drugs, they are good examples of how the Ro5 can <strong>be</strong> used as an alert<strong>,</strong> but should not <strong>necessarily</strong> be used as a filter.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>adapt fct name, if you change it according to this review </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>same here: fct name</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>One could move the first comment here to an extra markdown cell above the code. Same for the cell above.</p><p> </p><p>and maybe rephrase: </p><p>'# Concatenate molecules with Ro5 data'</p><p>-&gt; Concatenate dataframes containing molecule and Ro5 information</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Move this sentence below the next cell (where the fullfilled stats are shown)</p><p> </p><p>Add here instead:</p><p>Helper function to ... </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <ul><li><strong>thresholds of only 10, 5, and 5</strong><u>, respectively</u><strong>.</strong></li><li>we will scale all property values to an&nbsp;<strong>scaled threshold of 5. </strong>-&gt; maybe: ... scale them to fit in the <strong>range [0,5]?</strong></li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>no space in: ... radial (x-)axes </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Define input parameters that should stay the same <u>for</u> both radar charts:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Note also that drugs can also be administered via alternative routes, i.e. inhalation, skin penetration and injection.  </p><p>-&gt; Note <s>also</s> that drugs can also be administered via alternative routes, i.e. inhalation, skin penetration and injection.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

Comment (AndreaVolkamer) [inline teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>How can you plot information for&nbsp;an additional molecule in the&nbsp;radar charts that we were created in this talktorial?</p><p>-&gt; How can you plot information for&nbsp;an additional molecule in the&nbsp;radar charts that we <u>have</u> created in this talktorial?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/TeachOpenCADD/pull/49/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

**Comment by AndreaVolkamer** [comment]
@dominiquesydow the notebook looks great! 
I only suggested some minor textual changes. Great work!

**Comment by AndreaVolkamer** [comment]
and probably by accident this file also ended up in this PR?
teachopencadd/talktorials/004_compound_similarity/talktorial.ipynb

File: teachopencadd/talktorials/002_compound_adme/talktorial.ipynb (hunk 1 above)
Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>We will unify this paper reference note in the at the <code>base</code>  branch.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Done.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Done</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Agreed, that reads nicer!</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Done.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Thanks. Regarding the last point, I have deleted the comment on toxic molecules, since the focus is here on the metabolism.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Done.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Agreed.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Agreed.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Done.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>Thanks, that's better!</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>I removed the xtick labels, since we have a color code.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>I settled for <code>calculate_ro5_properties</code>  since we calculate not only the compliance but also the individual properties.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>We do not - we only use a scaling factor that will map 500 to 5 and 600 to 6. I could not find a name for this kind of scaling.</p>

Comment (dominiquesydow) [reply teachopencadd/talktorials/002_compound_adme/talktorial.ipynb:6]: <p>@t-kimber, maybe you have a better idea of how to phrase this? </p><p> </p>

**Comment by dominiquesydow** [comment]
> and probably by accident this file also ended up in this PR?
teachopencadd/talktorials/004_compound_similarity/talktorial.ipynb

True, thanks for spotting this.

**Comment by dominiquesydow** [comment]
@jaimergp and @AndreaVolkamer I think this is ready to be merged.

One open question could be addressed in a separate PR is needed: The phrasing for the threshold scaling that we do for the radar plot.

> The properties used for the Ro5 criteria are of different magnitudes. 
The MWT has a __threshold of 500__, whereas the number of HBAs and HBDs and the LogP have __thresholds of only 10, 5, and 5__, respectively. In order to visualize these different scales most simplistically, we will scale all property values to a __scaled threshold of 5__.
See @AndreaVolkamer's comment: https://github.com/volkamerlab/TeachOpenCADD/pull/49#discussion_r499580945

I tagged Talia if she has a nicer idea. It is up to you if you want to wait with merging.

**Comment by AndreaVolkamer** [comment]
> The properties used for the Ro5 criteria are of different magnitudes.
The MWT has a threshold of 500, whereas the number of HBAs and HBDs and the LogP have thresholds of only 10, 5, and 5, respectively. In order to visualize these different scales most simplistically, we will scale all property values to a scaled threshold of 5.
See @AndreaVolkamer's comment: #49 (comment)

Maybe we just make it explicit and name the respective ranges.
What is calculated is: [property value]/[property_threshold]*[scale_threshold]
* MW/500*5 = MW/100
* HBD/10*5=HDB/2
* HBA/5*5=HBA
* LogP/5*5=LogP

Or we simply say: 

> In order to visualize these different scales most simplistically, we scale all property values using the `name of helper fct` function. This results in a downscaling of the  MW by 100, HBD by 2, while HBA and LogP stay unchanged.

**Comment by dominiquesydow** [comment]
@AndreaVolkamer - thanks, that's great.

I updated the text as follows:

image

**Comment by dominiquesydow** [comment]
@jaimergp, I think this PR is ready to be merged then :)
